lib/ts/BufferWriter.h

```diff
@@ -0,0 +1,326 @@
+  virtual char *
+  auxBuffer()
+  {
+    return nullptr;
+  }
+
+  // Write the first n characters that have been placed in the auxiliary buffer.  This call invalidates the auxiliary buffer.
+  // This function should not be called if no auxiliary buffer is available.
+  //
+  virtual BufferWriter &
+  write(size_t n)
+  {
+    return *this;
+  }
+
+  // Returns number of total characters that can be written without causing an error condidtion.
+  //
+  virtual size_t capacity() const = 0;
+
+  // Total number of characters that have been written, including those discarded due to an error condition.
+  //
+  virtual size_t extent() const = 0;
+
+  // Total number of characters that are in the buffer (successfully written and not discarded).
+  //
+  size_t
+  size() const
+  {
+    size_t e = extent(), c = capacity();
+
+    return e < c ? e : c;
+  }
+
+  // Returns number of additional characters that can be written without causing an error condidtion.
+  //
+  size_t
+  remaining() const
+  {
+    return capacity() - size();
+  }
+
+  // Reduce the capacity by n characters, potentially creating an error condition.
+  //
+  virtual BufferWriter &clip(size_t n) = 0;
+
+  // If there is an error condition, this function clears it and sets the extent to the size.  It then increases the
+  // capacity by n characters.
+  //
+  virtual BufferWriter &extend(size_t n) = 0;
+
+  // Make destructor virtual.
+  //
+  virtual ~BufferWriter() {}
+};
+
+// A buffer writer that writes to an array of char that is external to the writer instance.
+//
+class FixedBufferWriter : public BufferWriter
+{
+protected:
+  FixedBufferWriter(char *buf, size_t capacity, size_t attempted) : _buf(buf), _capacity(capacity), _attempted(attempted) {}
+
+public:
+  // 'buf' is a pointer to the external array of char to write to.  'capacity' is the number of bytes in the array.
+  //
+  // If you create a instance of this class with capacity == 0 (and a nullptr buffer), you can use it to measure the number of
+  // characters a series of writes would result it (from the extent() value) without actually writing.
+  //
+  FixedBufferWriter(char *buf, size_t capacity) : FixedBufferWriter(buf, capacity, 0) {}
+
+  FixedBufferWriter &
+  write(char c) override
+  {
+    if (_attempted < _capacity) {
+      _buf[_attempted] = c;
+    }
+    ++_attempted;
+
+    return *this;
+  }
+
+  FixedBufferWriter &
+  write(const void *data, size_t length) override
+  {
+    size_t newSize = _attempted + length;
+
+    if (newSize <= _capacity) {
+      std::memcpy(_buf + _attempted, data, length);
+
+    } else if (_attempted < _capacity) {
+      std::memcpy(_buf + _attempted, data, _capacity - _attempted);
+    }
+    _attempted = newSize;
+
+    return *this;
+  }
+
+  // It's not clear to my why g++ needs this using declaration in order to consider the inherited versions of 'write' when
+  // resolving calls to a 'write' member ( wkaras@oath.com ).
+  //
+  using BufferWriter::write;
```

Comment:
I think this is a C++ spec issue and I think it was designed this way so that you could *not* pull in the superclass overloads.